activesupport/lib/active_support/message_verifier.rb

```diff
@@ -103,11 +103,17 @@ class MessageVerifier
 
     class InvalidSignature < StandardError; end
 
+    cattr_accessor :message_verifier_json_default_serializer, instance_accessor: false, default: false
+
     def initialize(secret, digest: nil, serializer: nil)
       raise ArgumentError, "Secret should not be nil." unless secret
       @secret = secret
       @digest = digest || "SHA1"
-      @serializer = serializer || Marshal
+      @serializer = serializer || if self.class.message_verifier_json_default_serializer
+        JSON
+      else
+        Marshal
+      end
```

Comment:
I believe we should force JSON here and leverage the configurable conditional elsewhere (explained below).

Reason being, we are focused on changing the default moving forward and less on backporting at the moment. 

For net new apps, they are free to provide their serializer of choice when they initialize a `MessageVerifier` so JSON as the default should be fine.
```ruby
@verifier = ActiveSupport::MessageVerifier.new('passwordlol', serializer: Marshal)
```

For existing apps that upgrade to Rails 7.0, if they were previously using the `MessageVerifier` with a default serializer, I think we're responsible for making sure that any existing payload returned to the app created prior to the upgrade should be accepted.

This could be accomplished with the following methods wrapping around the calls to `@serializer` in `MessageVerifier`:
```ruby
      def serialize(value)
        @serializer.dump(value)
      end

      def deserialize(value)
        @serializer.load(value)
      rescue ::JSON::ParserError
        if self.class.fallback_to_marshal_serialization
          Marshal.load(value)
        end
      end
```

Reply:
I don't hold this opinion very strongly, so I'm willing to change it, but I'm not sure I agree that this style is a better way to accomplish "we are focused on changing the default moving forward and less on backporting at the moment".

> For existing apps that upgrade to Rails 7.0, if they were previously using the MessageVerifier with a default serializer, I think we're responsible for making sure that any existing payload returned to the app created prior to the upgrade should be accepted.

As far as I understand both versions meet this criteria. Upgraded apps will not have the config value set changing the default serializer, so their usage of the default value will not change and will be Marshal, so existing payloads created prior to the upgrade are accepted.

---

I think the difference here is in what the intention of the configuration should be. We interpreted it as "turn this new feature on/off", the new feature being that the default is `JSON`, which I think the initialization method is a better match for.

With this the configuration is more an indication that "We no longer have a need for existing marshal payloads".

---

All that being said, as I mentioned at the beginning I am willing to change it if you disagree, what do you think?

Reply:
I had another thought, doesn't a change like this introduce a potential security issue for users that were *already* using a json-based serializer prior to rails 7.0?

Here is the scenario I'm picturing. 
`fallback_to_marshal_serialization` would need to default true for upgraded apps since the users who are using the default would need their workflow not to break.

You have a rails app on 6.x, you know about the evils of the Marshal serializer so you use a JSON-based one instead. 

Rails 7.0 comes out, you read the changlog and see an entry about the default being changed to JSON. You don't pay any more attention to that because you're already specifying a serializer. Something else in the upgrade path stops you from just turning on all the new features but you do upgrade to 7.0.

Now since `fallback_to_marshal_serialization` defaulted to true, you actually get this fallback behavior in your app even though you never used the marshal serializer in the first place. That means that an attacker could *send* you Marshal'd data and it would successfully decode, actually opening you up to the attack you were specifically mitigating by moving to JSON on your own.

Reply:
Wrapping the default in a config has been switch out with a hard default of JSON with a config based fallback